### PR TITLE
Update tutorial for Azure OpenAI

### PR DESCRIPTION
**Pull request by riedgar-ms**
Make the tutorial marginally less terse, and start work to incorporate Azure OpenAI endpoints

- Add an AzureOpenAI class for use with Azure OpenAI endpoints
- Tweak tutorial notebook

Contains significant contributions from @dfredriksenkbr

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -41,7 +41,8 @@ def find_version(*file_paths):
         "scipy",
         "aiohttp",
         "ordered_set",
-        "pyformlang"
+        "pyformlang",
+        "scipy"
```
Comment (riedgar-ms) [inline setup.py:45]: Note that if #469 is accepted, this will cease to be required.

**Comment by slundberg** [comment]
Thanks @riedgar-ms ! One high level question: Do we have to have a separate class for Azure OpenAI? I know in the past we had the OpenAI class support both Azure OpenAI and direct OpenAI. What forced the separation here?

(if we can keep them together then we can avoid some code duplication/drift)

Thanks again!

File: setup.py (hunk 1 above)
Comment (riedgar-ms) [reply setup.py]: I see that that PR was merged, so removing this.

**Comment by riedgar-ms** [comment]
@slundberg  and @dfredriksenkbr  I've just been doing some reworking, which has meant some fairly major refactoring work in `_openai.py` as well.

Unfortunately, the final cell in the example notebook I'm adding now ends with:
```
File /mnt/c/Users/riedgar/source/repos/guidance/guidance/models/_model.py:265, in Model.__add__(self, value)
    262         out = value(lm)
    264 # this flushes the display
--> 265 out._inplace_append("")
    267 return out

AttributeError: 'NoneType' object has no attribute '_inplace_append'
```
for reasons currently obscure - beyond that the end is obviously not quite being detected correctly. I don't have an OpenAI key, so I'm not sure if this is happening in the original notebook as well.

**Comment by riedgar-ms** [comment]
> Thank you for flagging that multiple-inheritance issue @dfredriksenkbr (perhaps the mixin should inherit from their bases instead?). I've also updated the notebook so that by default it's using AAD auth.
> 
> Tag @slundberg too

Changed the mixin to inherit. Since it's overriding methods, it seems logical to do this.

**Comment by slundberg** [comment]
Thanks @riedgar-ms  and @dfredriksenkbr ! I just fixed one merge issue (due to an incorrect inheritance fixed now in main). I have also been pushing bug fixes/improvements to the Remote class yesterday and today, so it might be worth a quick run through to see (I just started the unit tests again). We should also add some Azure specific unit tests :)

**Comment by slundberg** [comment]
hmmm: TypeError: Cannot create a consistent method resolution

**Comment by slundberg** [comment]
I think the issue here is that we should not have mixins, we should just have Azure inherit directly from OpenAI

**Comment by riedgar-ms** [comment]
Can you outline what class hierarchy you're expecting please @slundberg ?

**Comment by slundberg** [comment]
I pushed an update that I think fixes the python import errors (and OpenAI tests). I can't validate it for the AzureOpenAI endpoint though.

**Comment by riedgar-ms** [comment]
@slundberg I've added a smoke test of the chat class.

**Comment by slundberg** [comment]
@dfredriksenkbr thanks! I know that is the case for the openai endpoints, but I am not sure if that is different that AzureOpenAI. Do you know? If so we need to account for the difference.

**Comment by riedgar-ms** [comment]
@slundberg @dfredriksenkbr I've deployed a text-davinci-003 endpoint, and written an associated test. I've updated the Mixin with the change from `delta` to `text` as noted above. I don't have an OpenAI endpoint to test that change against, though (I would _hope_ that the actual 'call OpenAI' part is the same, but....)

**Comment by slundberg** [comment]
Alright, LTGM! Thanks both!

**Pull request merged**